### ISSUE TITLE
Should U+10000 to U+10FFFF be removed from Identifier Start and Part?

### ISSUE THREAD

**Issue by LongTengDao** [json5/json5#202]
Because that's not supported until ES6

**Comment by jordanbtucker** [comment]
Can you please provide a source that indicates that it isn't supported in ES5?

**Comment by jordanbtucker** [comment]
I see what you are saying, and you're right. It seems that ES5 treats [source text](https://www.ecma-international.org/ecma-262/5.1/#sec-6) as a sequence of Unicode code *units* (UTF-16) whereas ES6 treats [source text](https://www.ecma-international.org/ecma-262/6.0/#sec-source-text) as a sequence of Unicode code *points* (U+0000 to U+10FFFF).

For example, `𠮷` would be converted to UTF-16 using [Normalized Form C](https://unicode.org/reports/tr15/#Norm_Forms) as the two code units U+D842 and U+DFB7 in ES5. This is fine inside of strings, but is invalid inside of *IdentifierStart* because neither code unit belongs to the Unicode categories in *UnicodeLetter*. (They're not even valid Unicode code points because they're a pair of surrogates.)

However, `𠮷` is just seen as the code point U+20BB7 in ES6, which belongs to the Other Letter Unicode category, so it's valid in *IdentifierStart*.

Note that `\uD842\uDFB7` still fails to parse as *IdentifierStart* in ES5, ES6, and JSON5 because it's always treated as two code units.

I know you basically said all that. I just wanted to rephrase it to make it clearer to future readers.

---

As for JSON5, technically this goes deeper than just parsing *JSON5Identifier* since ES5 treats *all* source text as UTF-16. This implementation currently treats [source text as Unicode code points](https://github.com/json5/json5/blob/97e07f3a77070493a6400ba9f171fd4bfa606767/lib/parse.js#L87). However I think that *JSON5Identifier* is the only place that it really matters. Code points are automatically normalized inside strings, and there are no white space characters with code points larger than U+3000.

Since JSON5 is foremost a superset of JSON, and not a strict subset of ES5, and since it already violates the ES5 spec by [supporting U+2028 and U+2029 in strings](https://spec.json5.org/#separators) like JSON does, I think we should just keep this feature as another willful violation of the ES5 spec. However, the JSON5 spec should be updated to reflect this since it references ES5's *SourceCharacter* production.
